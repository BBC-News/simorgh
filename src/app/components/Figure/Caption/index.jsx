import styled from 'styled-components';
import { FF_NEWS_SANS_REG, C_STORM } from '../../../lib/constants/styles';
import mediaQuery from '../../../helpers/mediaQueries';

const Caption = styled.figcaption`
  background-color: #d5d0cd;
  color: ${C_STORM};
  font-family: ${FF_NEWS_SANS_REG};
  padding: 8px;

  // Font styling below is a subset of BBC GEL Typography "Long Primer"
  font-size: 0.9375em;
  line-height: 1.125em;
  ${mediaQuery.laptopAndLarger} {
    font-size: 0.875em;
  }
`;

export default Caption;
