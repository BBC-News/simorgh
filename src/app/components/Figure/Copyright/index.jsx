import React from 'react';
import { node } from 'prop-types';
import styled from 'styled-components';
import {
  FF_NEWS_SANS_REG,
  C_WHITE,
  C_EBON,
  GEL_SPACING,
} from '../../../lib/constants/styles';
import { T_MINION } from '../../../lib/constants/typography';
import VisuallyHiddenText from '../../VisuallyHiddenText';

const StyledCopyright = styled.span.attrs({
  role: 'text',
})`
  ${T_MINION};
  background-color: ${C_EBON};
  text-transform: uppercase;
  color: ${C_WHITE};
  padding: ${GEL_SPACING};
  font-family: ${FF_NEWS_SANS_REG};
  position: absolute;
  bottom: 0;
  right: 0;

  /*
    The hidden '.' is required for accessibility reasons
    https://github.com/bbc/simorgh/issues/456#issuecomment-419405600
  */
  &::after {
    content: '.';
    color: ${C_EBON}; /* This needs to match the background-color */
  }
`;

const Copyright = ({ children }) => (
  <StyledCopyright>
    <VisuallyHiddenText>Image source,</VisuallyHiddenText>
    {children}
  </StyledCopyright>
);

Copyright.propTypes = {
  children: node.isRequired,
};

export default Copyright;
