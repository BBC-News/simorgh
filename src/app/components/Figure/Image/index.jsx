import React from 'react';
import { number, string } from 'prop-types';
import StyledImg from './styledImg';

export { default as AmpImg } from './index.amp';

export const Img = ({ alt, height, src, srcset, width }) => {
  const props = { alt, src, height, width };

  if (srcset) {
    props.srcset = srcset;
  }

  return <StyledImg {...props} />;
};

Img.propTypes = {
  alt: string.isRequired,
  height: number.isRequired,
  src: string.isRequired,
  srcset: string,
  width: number.isRequired,
};

Img.defaultProps = {
  srcset: null,
};

const Image = Img;

export default Image;
