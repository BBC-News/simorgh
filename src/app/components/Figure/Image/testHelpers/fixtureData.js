export const landscape = {
  alt: 'Student sitting an exam',
  src:
    'https://ichef.bbci.co.uk/news/1024/cpsprodpb/7098/production/_104842882_students.jpg',
  width: 1024,
  height: 576,
};

export const portrait = {
  alt:
    'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.',
  src:
    'https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png',
  width: 640,
  height: 1138,
};

export const square = {
  alt: 'Tracks through the snow',
  src:
    'https://ichef.bbci.co.uk/news/1024/cpsprodpb/114FE/production/_104801907_79010.jpg',
  width: 1024,
  height: 1024,
};

export const custom = {
  alt: 'By Elisa Decker, from her series "Sidewalk"',
  src:
    'https://ichef.bbci.co.uk/news/445/cpsprodpb/164BB/production/_104032319_03270dcc-9dda-4bd4-96a0-db89f6b915ae.jpg',

  width: 445,
  height: 547,
};
