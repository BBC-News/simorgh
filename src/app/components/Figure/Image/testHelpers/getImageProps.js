function getImageProps(image, includeHeight) {
  const props = {
    alt: image.alt,
    src: image.src,
    width: image.width,
  };
  props.height = includeHeight ? image.height : null;
  return props;
}

export default getImageProps;
