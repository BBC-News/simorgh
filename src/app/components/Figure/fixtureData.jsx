import React from 'react';
import FigureContainer from '../../containers/Figure';

const imageAlt =
  'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.';
const imageSrc =
  'https://ichef.bbci.co.uk/news/640/cpsprodpb/A933/production/_101651334_bouquet_pa.jpg';
const imageRatio = 56.25;
const captionValue = 'This is a caption';
const captionValueContainingLink =
  'This is a caption [with a link](https://bbc.com/news)';
const copyrightText = 'Copyright Getty images';

const generateFixtureData = (caption, copyright) => (
  <FigureContainer
    src={imageSrc}
    alt={imageAlt}
    ratio={imageRatio}
    caption={caption}
    copyright={copyright}
  />
);

export const FigureImage = generateFixtureData();

export const FigureImageWithCaption = generateFixtureData(captionValue);

export const FigureImageWithCaptionContainingLink = generateFixtureData(
  captionValueContainingLink,
);

export const FigureImageWithCopyright = generateFixtureData(
  null,
  copyrightText,
);

export const FigureImageWithCopyrightAndCaption = generateFixtureData(
  captionValue,
  copyrightText,
);
