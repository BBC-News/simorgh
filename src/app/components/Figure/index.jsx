import React from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import LazyLoad from 'react-lazyload';
import Caption from './Caption';
import VisuallyHiddenText from '../VisuallyHiddenText';
import { GEL_SPACING_DBL } from '../../lib/constants/styles';

const copyrightOffscreenText = 'Copyright';

const renderCopyright = copyright => {
  const copyrightText = `${copyrightOffscreenText} ${copyright}`;
  return copyright !== 'BBC' ? (
    <VisuallyHiddenText>{copyrightText}</VisuallyHiddenText>
  ) : null;
};

const renderCaption = caption =>
  caption ? <Caption>{caption}</Caption> : null;

const StyledFigure = styled.figure`
  margin: 0;
  padding-bottom: ${GEL_SPACING_DBL};
  width: 100%;
`;

const Image = styled.img`
  display: block;
  width: 100%;
`;

const Figure = ({ src, alt, copyrightHolder, caption }) => (
  <StyledFigure>
    <LazyLoad height={200}>
      <Image src={src} alt={alt} />
      {renderCopyright(copyrightHolder)}
      {renderCaption(caption)}
    </LazyLoad>
  </StyledFigure>
);

Figure.propTypes = {
  alt: string.isRequired,
  src: string.isRequired,
  copyrightHolder: string.isRequired,
  caption: string,
};

Figure.defaultProps = {
  caption: null,
};

export default Figure;
