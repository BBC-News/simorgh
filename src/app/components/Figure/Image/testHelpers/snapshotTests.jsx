import React from 'react';
import { shouldMatchSnapshot } from '../../../../helpers/tests/testHelpers';
import { custom, landscape, portrait, square } from './fixtureData';

const snapshotTests = Component => {
  shouldMatchSnapshot(
    'should render landscape image correctly',
    <Component
      alt={landscape.alt}
      src={landscape.src}
      height={landscape.height}
      width={landscape.width}
    />,
  );
  shouldMatchSnapshot(
    'should render portrait image correctly',
    <Component
      alt={portrait.alt}
      src={portrait.src}
      height={portrait.height}
      width={portrait.width}
    />,
  );
  shouldMatchSnapshot(
    'should render square image correctly',
    <Component
      alt={square.alt}
      src={square.src}
      height={square.height}
      width={square.width}
    />,
  );
  shouldMatchSnapshot(
    'should render image with custom dimensions correctly',
    <Component
      alt={custom.alt}
      src={custom.src}
      height={custom.height}
      width={custom.width}
    />,
  );
  shouldMatchSnapshot(
    'should render image with srcset correctly',
    <Component
      alt={landscape.alt}
      src={landscape.src}
      srcset={landscape.srcset}
      height={landscape.height}
      width={landscape.width}
    />,
  );
};

export default snapshotTests;
