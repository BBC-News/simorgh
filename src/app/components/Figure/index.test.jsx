import React from 'react';
import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import Figure from './index';

const image = {
  alt:
    'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.',
  src:
    'https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png',
};

const caption = 'This is a caption!';

describe('Figure', () => {
  describe('with a caption', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure {...image} copyrightHolder="BBC" caption={caption} />,
    );
  });

  describe('without a caption', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure {...image} copyrightHolder="BBC" />,
    );
  });

  describe('with non BBC copyright', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure {...image} copyrightHolder="Getty images" />,
    );
  });
});
