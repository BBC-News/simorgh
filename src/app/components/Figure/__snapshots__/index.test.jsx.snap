// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure with a caption should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c2 {
  display: block;
  width: 100%;
}

.c4 {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  top: 0;
}

.c3 {
  background-color: #D5D0CD;
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding: 0.5rem;
  width: 100%;
}

.c1 {
  position: relative;
}

@media (min-width:37.5em) {
  .c3 {
    padding: 0.5rem 1rem;
  }
}

<figure
  className="c0"
>
  <div
    className="c1"
  >
    <img
      alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
      className="c2"
      src="https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg"
    />
  </div>
  <figcaption
    className="c3"
  >
    <span
      className="c4"
    >
      Image caption, 
    </span>
    This is a caption
  </figcaption>
</figure>
`;

exports[`Figure with caption and non-BBC copyright should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c2 {
  display: block;
  width: 100%;
}

.c5 {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  top: 0;
}

.c4 {
  background-color: #D5D0CD;
  color: #404040;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  padding: 0.5rem;
  width: 100%;
}

.c3 {
  font-size: 0.75em;
  line-height: 1em;
  background-color: #222222;
  color: #FFFFFF;
  padding: 0.5rem;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  position: absolute;
  bottom: 0;
  right: 0;
}

.c1 {
  position: relative;
}

@media (min-width:37.5em) {
  .c4 {
    padding: 0.5rem 1rem;
  }
}

<figure
  className="c0"
>
  <div
    className="c1"
  >
    <img
      alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
      className="c2"
      src="https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg"
    />
    <span
      className="c3"
    >
      © 
      Getty Images
    </span>
  </div>
  <figcaption
    className="c4"
  >
    <span
      className="c5"
    >
      Image caption, 
    </span>
    This is a caption
  </figcaption>
</figure>
`;

exports[`Figure with non-BBC copyright should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c2 {
  display: block;
  width: 100%;
}

.c3 {
  font-size: 0.75em;
  line-height: 1em;
  background-color: #222222;
  color: #FFFFFF;
  padding: 0.5rem;
  font-family: ReithSansNewsRegular,Helvetica,Arial,sans-serif;
  position: absolute;
  bottom: 0;
  right: 0;
}

.c1 {
  position: relative;
}

<figure
  className="c0"
>
  <div
    className="c1"
  >
    <img
      alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
      className="c2"
      src="https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg"
    />
    <span
      className="c3"
    >
      © 
      Getty Images
    </span>
  </div>
</figure>
`;

exports[`Figure without a caption should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c2 {
  display: block;
  width: 100%;
}

.c1 {
  position: relative;
}

<figure
  className="c0"
>
  <div
    className="c1"
  >
    <img
      alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
      className="c2"
      src="https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg"
    />
  </div>
</figure>
`;
